### PR TITLE
Feature/caching

### PR DESCRIPTION
**Pull request by worksofhart**
# Description

Moved caching to API call layer. Improved error handling

Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Change Status

- [ ] Complete, tested, ready to review and merge
- [x] Complete, but not tested (may need new tests)
- [ ] Incomplete/work-in-progress, PR is for discussion/feedback

# How Has This Been Tested?

- [ ] Test A
- [ ] Test B

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] My code has been reviewed by at least one peer
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] There are no merge conflicts

### CODE REVIEW

File: src/components/Grid/Grid.jsx [hunk 1]
```diff
@@ -46,54 +46,45 @@ const Grid = () => {
...
+      ? deserialize(localStorage.getItem('dates'))
+      : []
+  )
 
   const [spinner, setSpinner] = useState(false)
   const [token] = useGetToken()
   const [exportCSV, setExportCSV] = useState(null)
 
   useEffect(() => {
+    setSpinner('One moment please...')
```
Comment (AdnanCodes) [inline src/components/Grid/Grid.jsx]: @worksofhart setSpinner accepts true or false to determine loading spinner.

File: src/components/Grid/Grid.jsx [hunk 2]
```diff
@@ -142,38 +133,107 @@ const Grid = () => {
...
   function datesHandler(dates) {
-    setDateRanges(dates)
-    localStorage.setItem('dates', JSON.stringify(dates))
+    if (dates) {
+      setDateRanges(dates)
+      localStorage.setItem('dates', serialize(dates))
+    } else localStorage.removeItem('dates')
   }
 
   function restoreQuery() {
```
Comment (AdnanCodes) [inline src/components/Grid/Grid.jsx]: This makes a lot easier to understand the caching. Incredible work.

File: src/components/Grid/Grid.jsx [hunk 3]
```diff
@@ -142,38 +133,107 @@ const Grid = () => {
...
+      dropdownHandler(currency, setCurrency, null, 'cur')
+      datesHandler(
+        startDate && endDate ? [moment(startDate), moment(endDate)] : []
+      )
+    }
   }
 
   function resetSearch() {
+    setErr(false)
+    setSpinner('One moment please...')
```
Comment (AdnanCodes) [inline src/components/Grid/Grid.jsx]: Same goes here for setSpinner

File: src/components/Grid/Grid.jsx [hunk 4]
```diff
@@ -142,38 +133,107 @@ const Grid = () => {
         queryUpdater(null)
       }
     }
-    localStorage.setItem(prefix, JSON.stringify(value))
+    // localStorage.setItem(prefix, JSON.stringify(value))
+  }
+
+  function serialize(collection) {
```
Comment (rhokath) [inline src/components/Grid/Grid.jsx]: what's this function doing?

File: src/components/Grid/Grid.jsx [hunk 5]
```diff
@@ -197,25 +257,20 @@ const Grid = () => {
...
-          baseURL:
-            process.env.NODE_ENV !== 'development'
-              ? 'https://sauti-africa-market-master.herokuapp.com/'
-              : 'http://localhost:8888/'
-        }
-      )
+      .get(query)
       .then(async res => {
-        localStorage.setItem('rowdata', JSON.stringify(res.data.records))
+        if (res.error) throw new Error(res.error)
```
Comment (rhokath) [inline src/components/Grid/Grid.jsx]: cool

File: src/components/Grid/Grid.jsx [hunk 6]
```diff
@@ -284,18 +334,19 @@ const Grid = () => {
...
         : ''
-
-    axiosWithAuth([token])
-      .get(
-        `http://localhost:8888/sauti/client/export/?currency=${currency ||
-          'USD'}${countryQuery || ''}${marketQuery || ''}${pCatQuery ||
-          ''}${pAggQuery || ''}${productQuery || ''}${dateRangeQuery}`
-      )
+    setErr(false)
+    const query = `http://localhost:8888/sauti/client/export/?currency=${currency ||
```
Comment (rhokath) [inline src/components/Grid/Grid.jsx]: should we change this to the master/staging url?

Comment (AdnanCodes) [reply src/components/Grid/Grid.jsx]: good catch and I already pulled it staging, so make commit to staging instead

File: src/components/SideNav/SideNav.jsx [hunk 7]
```diff
@@ -545,35 +545,72 @@ const SideNav = () => {
...
+                    </span>
+                  </Table.Cell>
                   <Table.Cell>OK - Worked as expected!</Table.Cell>
                 </Table.Row>
                 <Table.Row>
-                  <Table.Cell>400 🙃</Table.Cell>
-                  <Table.Cell>Bad Request -  Request was missing a parameter.</Table.Cell>
+                  <Table.Cell>
+                    400{' '}
+                    <span role="img" aria-label="upside-down smiley face">
```
Comment (rhokath) [inline src/components/SideNav/SideNav.jsx]: I appreciate you!

File: src/utils/axiosWithAuth.js [hunk 8]
```diff
@@ -1,9 +1,63 @@
 import axios from 'axios'
 
-export const axiosWithAuth = token => {
-  return axios.create({
-    headers: {
-      Authorization: `Bearer ${token}`
+let cache = {
```
Comment (rhokath) [inline src/utils/axiosWithAuth.js]: you are a wizard

File: src/utils/axiosWithAuth.js [hunk 9]
```diff
@@ -1,9 +1,63 @@
...
+      }
+      this.cacheData.set(key, value)
+    }
+  },
+  del: function(key) {
+    this.cacheData.delete(key)
+  }
+}
+
+export const axiosWithAuth = (token, exclude) => {
```
Comment (rhokath) [inline src/utils/axiosWithAuth.js]: axiosWithAuth on steroids, I like it

**Pull request merged**